public/app/plugins/datasource/tempo/datasource.ts

```diff
@@ -1,121 +1,63 @@
-import { serializeParams } from 'app/core/utils/fetch';
-
 export type TempoQuery = {
   query: string;
 } & DataQuery;
 
-export class TempoDatasource extends DataSourceApi<TempoQuery> {
-  constructor(private instanceSettings: DataSourceInstanceSettings, private readonly timeSrv: TimeSrv = getTimeSrv()) {
+export class TempoDatasource extends DataSourceWithBackend<TempoQuery> {
+  constructor(instanceSettings: DataSourceInstanceSettings) {
     super(instanceSettings);
   }
 
-  async metadataRequest(url: string, params?: Record<string, any>): Promise<any> {
-    const res = await this._request(url, params, { hideFromInspector: true }).toPromise();
-    return res.data.data;
-  }
-
   query(options: DataQueryRequest<TempoQuery>): Observable<DataQueryResponse> {
-    // At this moment we expect only one target. In case we somehow change the UI to be able to show multiple
-    // traces at one we need to change this.
-    const id = options.targets[0]?.query;
-    if (id) {
-      return this._request(`/api/traces/${encodeURIComponent(id)}`).pipe(
-        map((response) => {
-          return {
-            data: [
-              new MutableDataFrame({
-                fields: [
-                  {
-                    name: 'trace',
-                    type: FieldType.trace,
-                    values: response?.data?.data || [],
-                  },
-                ],
-                meta: {
-                  preferredVisualisationType: 'trace',
+    return super.query(options).pipe(
+      map((response) => {
+        if (response.error) {
+          return response;
+        }
+
+        return {
+          data: [
+            new MutableDataFrame({
+              fields: [
+                {
+                  name: 'trace',
+                  type: FieldType.trace,
+                  values: [JSON.parse((response.data as DataFrame[])[0].fields[0].values.get(0))],
                 },
-              }),
-            ],
-          };
-        })
-      );
-    } else {
-      return of({
-        data: [
-          new MutableDataFrame({
-            fields: [
-              {
-                name: 'trace',
-                type: FieldType.trace,
-                values: [],
+              ],
+              meta: {
+                preferredVisualisationType: 'trace',
               },
-            ],
-            meta: {
-              preferredVisualisationType: 'trace',
-            },
-          }),
-        ],
-      });
-    }
+            }),
+          ],
+        };
+      })
+    );
   }
 
   async testDatasource(): Promise<any> {
-    try {
-      await this._request(`/api/traces/random`).toPromise();
-    } catch (e) {
-      // If all went well this request will get back with 400 - Bad request
-      if (e?.status !== 400) {
-        throw e;
-      }
+    const response = await super.query({ targets: [{ query: '', refId: 'A' }] } as any).toPromise();
```

Comment:
I would like notice that this tests of datasources does not verify datasource really. Backends might fail with "404 Not Found", but tests pass and user does not have idea why and is not aware of misconfiguration until it attempt to query.

Reply:
I have found misconfiguration on my side which push me to look on that code. I provide URL like `http://tempo:3100` instead of `http://tempo:3100/tempo`. This looks like something - at least - to catch during testing of configuration.

Reply:
Hmm yeah this is not great test but also not sure if there is any API that would give us definite "working" status in Tempo. Can you create a separate issue for this?